### PR TITLE
Why Kubernetes is the best technology for running Cloud-Native Database

### LINKED ISSUE #2544: Why Kubernetes is the best technology for running Cloud-Native Database
### Proposed title of article
Why Kubernetes is the best technology for running Cloud-Native Database

### Introduction:
The discussion about migrating workloads to the cloud has been there for a long time, but a looking at the application portfolios of many IT organizations demonstrates that there’s still a lot of work to be done. In many cases, challenges with persisting and moving data in clouds continue to be the key limiting factor slowing cloud adoption, despite the fact that databases in the cloud have been available for years.  
For this reason, there has been a surge of recent interest in a data infrastructure that is designed to take maximum advantage of the benefits that cloud computing provides. A good example of a cloud-native database is the K8ssandra project. It packages Apache Cassandra and supporting tools into a production-ready Kubernetes deployment. 
This raises an interesting question: must a database running on Kubernetes be considered cloud-native? This article will dive into the issue and try to come up with the answer.

### Key takeaways: 
In this article the reader will able to:
- understand the concept of Cloud-Native Databases
- explore different Kubernetes design principles
- draw the conclusion on the best designs based on the principles discussed earlier

### References: 
N/A

### PR DESCRIPTION
**Pull request by ephraa**
### Article checklist

This fixes #2544.

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by ephraa** [comment]
This solves #2544

**Comment by hectorkambow** [comment]
@ephraa 
Good afternoon and thank you for submitting your article for publishing. It has been entered into our queue and should be reviewed (and published pending any changes) as soon as a peer reviewer is finished with the ones in the queue before it. 🚀

**Comment by hectorkambow** [comment]
plagiarism check done

File: content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+﻿### Introduction
```
Comment (mbaabuones) [inline content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md:1]:
```suggestion
For a long time, every organization’s dream is to migrate its workload to the cloud. However, the statistics from most IT organizations still show that there is still much work to be done. 

Although databases in the cloud have been around for a long time, the challenges with persisting and moving the cloud data have led to slow cloud adoption. For this reason, the demand for data infrastructure that takes full advantage of cloud computing has been on the rise.

This article provides an overview of cloud native databases and explains why Kubernetes is the best technology for running them.

File: content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+For a long time, every organization’s dream is to migrate its workload to the cloud. However, the statistics from most IT organizations still show that there is still much work to be done.
+
+Although databases in the cloud have been around for a long time, the challenges with persisting and moving the cloud data have led to slow cloud adoption.
+
+For this reason, the demand for data infrastructure that takes full advantage of cloud computing has been on the rise.
+
```
Comment (mbaabuones) [inline content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md:7]: After the above suggested paragraphs, add a new sub-title 'An overview of cloud native databases'. Under this sub-title, provide the definition of a cloud native database and add the other content.

File: content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+A simple confusion arises if it is a must that any database running on Kubernetes be considered cloud-native. [Kubernetes](https://kubernetes.io/) was initially developed for [stateless workloads](https://dzone.com/articles/kubernetes-and-running-stateful-workloads), but with recent developments such as persistent volumes and stateful sets, it is now possible to run [stateful workloads](https://dzone.com/articles/kubernetes-and-running-stateful-workloads).
+
+The longtime DevOps practitioners who were at first skeptical about running databases on Kubernetes have now started adopting it, and best practices are emerging.
+
+For the databases to be more cloud-native as they should be, there is a great need to embrace what Kubernetes brings to the table.
+
+A truly cloud-native approach has to be followed, meaning adopting vital elements of the Kubernetes design paradigm. For example, a database that can work effectively on Kubernetes can be considered cloud-native.
+
+### Kubernetes Design Principles
```
Comment (mbaabuones) [inline content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md:18]: Connect these principles to the main objective of the article. For example, you can provide the following statement, 'Lets look at the design principles of Kubernetes to understand why it is the best technology for running cloud native databases.'

File: content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+In Kassandra, for every `CassandraDataCenter` resource, the `CQL` access should be exposed as a service. The APIs for metrics and management is accessed by the `cass-operator` and the [Prometheus Service Monitor](https://medium.com/kubernetes-tutorials/simple-management-of-prometheus-monitoring-pipeline-with-the-prometheus-operator-b445da0e0d1a) in every Cassandra node.
+
+Kubernetes comes with features that allow sharing of encryption keys and the configuration of administrative accounts. For example, most K8ssandra deployments prefer changing Cassandra's default admin account details with a new admin username and password.
+
+#### Principle 5: Preferring declarative configuration
+With the declarative approach in Kubernetes, the desired state of resources and controllers are specified. It assists in the manipulation of the underlying infrastructure to achieve the specified state.
+
+The `cass-operator` specifies the required number of nodes in a given cluster and helps manage the details of adding new nodes for scaling up and choosing which nodes to remove to scale down.
```
Comment (mbaabuones) [inline content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md:58]: simplify this sentence.

File: content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+The next phase of operators ought to set the rules for stored data size, the number of transactions per second, or even both.
+
+In the future, it may be possible to specify the cluster sizes and move less frequently needed data to object storage.
+
+### Conclusion
+It is clear enough that Kubernetes provides the best practices for cloud-native database implementations, and innovation is still ongoing.
+
+Solutions for integrating Kubernetes clusters are still in the development phase, but in the future, it will be easy to manage multi-datacenter Cassandra clusters in Kubernetes.
+
+Cassandra community is constantly working to make more extensions for management and metrics to be part of the core Apache project so that Cassandra will be more cloud-native to everyone and an out-of-the-box solution.
```
Comment (mbaabuones) [inline content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md:69]: Simplify this sentence. Tip- avoid writing very long sentences.

File: content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+#### Principle 4: Making the configuration secure
+Networking in Kubernetes should be ensured to be secure all the time; ports have to be precisely exposed to be accessed externally to a pod. It sets an important criterion when deploying the database on how every control plane and data plane interface is exposed, also the interfaces that should be exposed through the Kubernetes service.
+
+In Kassandra, for every `CassandraDataCenter` resource, the `CQL` access should be exposed as a service. The APIs for metrics and management is accessed by the `cass-operator` and the [Prometheus Service Monitor](https://medium.com/kubernetes-tutorials/simple-management-of-prometheus-monitoring-pipeline-with-the-prometheus-operator-b445da0e0d1a) in every Cassandra node.
```
Comment (mbaabuones) [inline content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md:51]: The APIs for metrics and management are accessed

Comment (mbaabuones) [inline content/articles/why-kubernetes-is-the-best-technology-for-running-cloud-native-database/index.md:51]: K8ssandra

**Review by mbaabuones — CHANGES_REQUESTED** [review]
@ephraa Great first article! The article is quite informative. I am glad to welcome you to the Section Engineering Program. I am happy that you have already uploaded your hero image, photo and author bio. That is commendable.

In your subsequent PRs, ensure you have a statement that explains what your article covers (see the comments for more details). Please work on the minor changes that I have suggested to you. Ping me when you are done.

**Comment by ephraa** [comment]
@mbaabuones The suggestions and points made have been well noted. The changes requested have been made, kindly review them.
Thank you.

**Review by mbaabuones — APPROVED** [review]

**Comment by mbaabuones** [comment]
@ephraa Good job. Please note the changes I made after the second paragraph. In your subsequent PRs, ensure you have a statement that explains what your article covers.

**Comment by ephraa** [comment]
@mbaabuones I have noted the changes and suggestions made, I will pay more attention to future articles.
Thank you

**Pull request merged**